### DISCUSSION TITLE
Multiple (write) sessions in duckDB

### DISCUSSION THREAD

**Discussion by avico78** [duckdb/duckdb#3379] [General]
Last few days , Im working with the duckDB and I really amazed how good it is ,
My dataset is multiple large data source coming from files, Tables(Oracle)
which I need to preform heavy join queries ,
I need to be able to store some of the data so in next run i will not need to start from beginning ,
My program run in different threads where the idea is to minimize full data scan  and gain better performance ,
so in example the steps would be something : 

read x.csv (1GB)
read table a(1GB)
--> preform join between --> save to _result1

read y.csv(1GB) 
read table b(1GB)
load _result1 to memory
--> preform join between the 3 --> save to _result2

and so on ...

My doubt : 
DuckDB provide great performance once data is loaded to tables,
but from other hand , i couldn't find how I can load the data in parallel ,
for in-memory - I guess no issue but as the program run in threads ,
each thread doesn't see the other data and i can share the objects ,
if I work in "file mode" -and save all to table I'm getting error  for file locked/in use ,

any suggestion ?

**Comment by Alex-Monahan** [comment]
Could you provide us a few more details (what language, and maybe your code?)? Ideally, could you build a small, reproducible code example?

**Reply by avico78** [reply]
Thanks @Alex-Monahan 

Writing in Python 3.8.5/3.9
My program is much more complex as the core functionality is a task orchestrator,
so it can execute tasks by dependencies + it does data pipeline between the tasks (optional) ,
of course ,it use multithreading for executing parallel tasks.
```
           task2 -- task3 
task1 /                          \
       \   task4   -- task5  -- task7
                  \                /
                      task6
````
in the context of duckdb : 
task1 --> prepare the data  and save it disk as DF (pickly,parquet)
task2 --> load file to duckdb and save it as table1 
task3 --> duckdb query the data and save to new table2
task4--> load the data to duckdb and save as table3
task5--> query table3 and save on duckdb table4
task6-->query table3 and save on duckdb table5
task7 --> query join table5,table4,table2

At beginning , I start with REDIS but it lack of  sql functionalities which duckdb does ,
of course also as is more oriented to data analysis thanks to the sql functionalities it provides,

The things that i need to overcome:
I wish i could save the processed data directly on the duckdb (file),
and reuse data saved table so in next task/s (there's might be on different thread (see task1->task6).
this obviously immediately failed as it wont allow multiple threads accessing same file: 

```
RuntimeError: IO Error: Cannot open file "mydb.duckdb": The process cannot access the file because it is being used by another process.
````

Second option would work in-memory ,
where the problem ,the session lives per thread and once exist ,data is not longer avilable.

in example , if load data to memory in task1 , in task6 it no longer available .
basically , I can save each of duckdb process (in example query the data)  to file(pickle) 
and load it anywhere i need ,in other worde will use duckdb as query engine and thats it .

that really make things little complex and will waste time on writing data to disk and load it back when needed.

I wrote some fast and dirty example which hopefully explain the situation :  

```python
import pandas as pd
import threading
import duckdb
import os


duck_conn_string = ''

# duck_conn = duckdb.connect(duck_conn_string)
# duck_cursor = duck_conn.cursor()


def task2(df,table_name):
        print("ID of process running task 2: {}".format(os.getpid()))
        duck_conn = duckdb.connect(duck_conn_string)
        duck_cursor = duck_conn.cursor()
        duck_cursor.execute(f"CREATE TABLE {table_name}  as select * from {df.name} ")
	# print("Square: {}".format(num * num))

def task1(df,table_name):
        duck_conn = duckdb.connect(duck_conn_string)
        duck_cursor = duck_conn.cursor()        
        duck_cursor.execute(f"CREATE TABLE {table_name}  as select * from {df.name} ")
        #Table exist
        print(duck_cursor.execute(f"select count(*) from {table_name}").df())

def count1(table_name):
        
        duck_conn = duckdb.connect(duck_conn_string)
        duck_cursor = duck_conn.cursor()
        print(duck_cursor.execute(f"select count(*) from {table_name}").df())
        

if __name__ == "__main__":
        #create some data
        dummy_data1 = {
        'id': ['1', '2', '3', '4', '5'],
        'Feature1': ['A', 'C', 'E', 'G', 'I'],
        'Feature2': ['B', 'D', 'F', 'H', 'J']}
        df1 = pd.DataFrame(dummy_data1, columns = ['id', 'Feature1', 'Feature2'])
        df2 = pd.DataFrame(dummy_data1, columns = ['id', 'Feature1', 'Feature2'])
        df1.name = "df1"
        df2.name = "df2"
	# creating first thread for creating task1 table
        t1 = threading.Thread(target=task1, args=(df1,"table1"))
        # creating second thread for creating task1 table
        t2 = threading.Thread(target=task2, args=(df2,"table2"))

	# starting thread 1
        t1.start()
	# starting thread 2
        t2.start()

	# wait until thread 1 is completely executed
        t1.join()

        #This part will return an error tablel1 doesn't exist as table1 created in separated thread 
        #
        # duck_conn = duckdb.connect(duck_conn_string)
        # duck_cursor = duck_conn.cursor()
        # t3 = threading.Thread(target=count1, args=(duck_cursor,"table1",))
        # t3.start()
	# wait until thread 2 is completely executed
        t2.join()
        # t3.join()
	# both threads completely executed
        print("Done!")


````


So my question , how i can still use duckdb but still will work with threads and make data available ,
or from other hand , save all processed data to pickle and load everywhere i need.

Would be thanks for any advise/tip ,

**Comment by avico78** [comment]
anyone?
